File: 0242-valid-anagram/0242-valid-anagram.java
class Solution {
    public boolean isAnagram(String s, String t) {

//conert both to lowercase to ignore case match
        s = s.toLowerCase();
        t = t.toLowerCase();
//strip of all the white spaces
        s = s.replace(" ","");
        t = t.replace(" ","");
//initialize the bucket array
         int[] counts = new int[26];
// fill in the buckets
         for (int i=0; i<s.length();i++){
             counts[s.charAt(i)-'a']++;
         }                             
//empty the buckets
          for (int i=0; i<t.length();i++){
             counts[t.charAt(i)-'a']--;
         }                                          
 //check if all buckets are empty
          for (int count:counts){
              if(count!=0)
              return false;    
          }                                        
                                                  
         return true;                                        
        
        
    }
}

File: 0242-valid-anagram/0242-valid-anagram.py
#optimized approach
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # Convert both strings to lowercase to ignore case match
        s = s.lower()
        t = t.lower()

        # Strip off all the white spaces
        s = s.replace(" ", "")
        t = t.replace(" ", "")

        # Initialize the bucket array
        counts = [0] * 26

        # Fill in the buckets
        for char in s:
            counts[ord(char) - ord('a')] += 1 #ord(char) is used to get the ASCII value of the character char. 

        # Empty the buckets
        for char in t:
            counts[ord(char) - ord('a')] -= 1

        # Check if all buckets are empty
        for count in counts:
            if count != 0:
                return False

        return True
